core/trino-main/src/test/java/io/trino/operator/TestExchangeClient.java

```diff
@@ -400,6 +400,104 @@ public void testStreamingAddLocation()
+
+        assertPageEquals(getNextPage(exchangeClient), createPage(1));
+
+        processor.setComplete(location1);
+
+        assertFalse(tryGetFutureValue(exchangeClient.isBlocked(), 10, MILLISECONDS).isPresent());
+
+        RuntimeException randomException = new RuntimeException("randomfailure");
+        processor.setFailed(location2, randomException);
+
+        assertThatThrownBy(() -> getNextPage(exchangeClient)).hasMessageContaining("Encountered too many errors talking to a worker node");
+
+        assertFalse(exchangeClient.isFinished());
+    }
+
+    @Test
+    public void testDeduplicationTaskFailure()
+            throws Exception
+    {
+        DataSize maxResponseSize = DataSize.of(10, Unit.MEGABYTE);
+        MockExchangeRequestProcessor processor = new MockExchangeRequestProcessor(maxResponseSize);
+
+        TaskId attempt0Task1 = new TaskId(new StageId("query", 1), 0, 0);
+        TaskId attempt1Task1 = new TaskId(new StageId("query", 1), 1, 0);
+        TaskId attempt1Task2 = new TaskId(new StageId("query", 1), 2, 0);
+
+        URI attempt0Task1Location = URI.create("http://localhost:8080/1/0");
+        URI attempt1Task1Location = URI.create("http://localhost:8080/1/1");
+        URI attempt1Task2Location = URI.create("http://localhost:8080/2/1");
+
+        processor.setFailed(attempt0Task1Location, new RuntimeException("randomfailure"));
+        processor.addPage(attempt1Task1Location, createPage(1));
+        processor.setComplete(attempt1Task1Location);
+        processor.setFailed(attempt1Task2Location, new RuntimeException("randomfailure"));
+
+        ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(scheduler, DataSize.of(1, Unit.MEGABYTE), RetryPolicy.QUERY);
+
+        ExchangeClient exchangeClient = new ExchangeClient(
+                "localhost",
+                DataIntegrityVerification.ABORT,
+                buffer,
+                maxResponseSize,
+                1,
+                new Duration(1, SECONDS),
+                true,
+                new TestingHttpClient(processor, scheduler),
+                scheduler,
+                new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext(), "test"),
+                pageBufferClientCallbackExecutor,
+                (taskId, failure) -> {});
+
+        exchangeClient.addLocation(attempt0Task1, attempt0Task1Location);
+        assertFalse(tryGetFutureValue(exchangeClient.isBlocked(), 10, MILLISECONDS).isPresent());
+
+        exchangeClient.addLocation(attempt1Task1, attempt1Task1Location);
+        exchangeClient.addLocation(attempt1Task2, attempt1Task2Location);
+        assertFalse(tryGetFutureValue(exchangeClient.isBlocked(), 10, MILLISECONDS).isPresent());
+
+        exchangeClient.noMoreLocations();
+        exchangeClient.isBlocked().get(10, SECONDS);
+
+        assertThatThrownBy(() -> getNextPage(exchangeClient)).hasMessageContaining("Encountered too many errors talking to a worker node");
```

Comment:
Why does this cause an "Encountered too many errors ..." error?

Reply:
A failure injected at the line 468 (`processor.setFailed(attempt1Task2Location, new RuntimeException("randomfailure"));`)

Reply:
Yes, but why "too many errors" instead of "randomfailure"?

Reply:
The `RuntimeException("randomfailure")` is injected at the HTTP request processing level, so the original error is getting lost